### ISSUE TITLE
Closing Connection for one FileHandle, closes Share for every other.

### ISSUE THREAD

**Issue by TheRealHypo** [hierynomus/smbj#395]
So I wrote a Test today to see how fast we can merge our Application from SmbFile (Jcifs) to this one here. 

But I now got some Problems while opening multiple file handles at once, and doing different tasks with each. 

e.g. I open 4 different Files in the same folder, with a GENERIC_ALL mask. no problems there. 

I write some stuff to one File, read other stuff from other File, delete another one and then close the first handle. In my Case I have a new Object, containing Connection, Session and DiskShare, and made a close() method that closes all of them. 

But when I close the Connection object on a Single fileHandle, it closes the DiskShare on all other opened Files as well. Is that desired behaviour ? 

No, Connection is not a static variable, I checked that a hundred times by now. 

My only option now is it to make Connection static, and keep it alive throughout the whole application run, Which I wanted to avoid. 

Any Tips here ? Code is in Groovy, I hope that is not a problem.

The following Test fails on each Consecutive 'close()' because it seems to be already closed.

    @Test
    void listFilesInFolderReturnFiles() {
        SmbFile a = new SmbFile("01_Content/05_TeamBar/TestCase/log.txt").open()
        SmbFile b = new SmbFile("01_Content/05_TeamBar/TestCase/xxxxxxx.jar").open()
        SmbFile c = new SmbFile("01_Content/05_TeamBar/TestCase/Scrxxxxuite").open()
        SmbFile d = new SmbFile("01_Content/05_TeamBar/TestCase/xxxxxx-xxxx.jar").open()

        List<SmbFile> expected = [a,b,c,d]

        SmbFile file = new SmbFile("01_Content/05_TeamBar/TestCase/").open()
        List<SmbFile> actual = file.listFiles(null, false)*.open()
        assertEquals("All files should be opened", expected, actual)

        // '*' in Groovy Code applies a method to all List elements.
        // First .close() is successful, each that follows fails. In this case no Problem.
        // but If I wanted to do stuff with other files it would get bad.
        expected*.close()
        actual*.close()
    }

Below is a small part of the Wrapping Code I use.

    private DiskEntry diskEntry
    private DiskShare share

    // DEFAULT Meta
    private EnumSet accessMask
    private EnumSet fileAttributes
    private EnumSet shareAccess
    private SMB2CreateDisposition createDisposition
    private EnumSet createOptions

    /**
     * Creates a new SmbFile reference at the given path.
     *
     * The path has to be relative to the Share name on the FileServer.
     * For example given the following Smb URI:
     *      smb://xxxxxx/xxxxxx/01_Content/05_TeamBar/TestCase/
     *
     * The path part would be:
     *      01_Content/05_TeamBar/TestCase/
     *
     * Though not extendedly tested yet, please try to use URI's that end with a '/' for directories.
     *
     * The Default Metadata is:
     *      AccessMask: GENERIC_ALL
     *      FileAttributes: NORMAL
     *      ShareAccess: ALL (Read, Write, Delete)
     *      Disposition: FILE_OPEN
     *      Options: None
     *
     * @param path relative path on fileserver
     */
    SmbFile(String path) {
        this.path = sanitizePath(path)
        accessMask = EnumSet.of(AccessMask.GENERIC_ALL)
        fileAttributes = EnumSet.of(FileAttributes.FILE_ATTRIBUTE_NORMAL)
        shareAccess = EnumSet.allOf(SMB2ShareAccess.class)
        createDisposition = SMB2CreateDisposition.FILE_OPEN
        createOptions = EnumSet.noneOf(SMB2CreateOptions.class)
        share_name = xxxx something from Properties xxxx
    }

    /**
     * Connects with the remote Server and creates a Session for this file reference.
     *
     * Note: This will NOT open the File !
     *
     */
    private void openSmbSession() {
        if (connection == null) {
            connection = client.connect(HOST)
        }
        createSession()
    }

    private void createSession() {
        AuthenticationContext ac = new AuthenticationContext(USERNAME, PASSWORD.toCharArray(), DOMAIN)
        session = connection.authenticate(ac)
        share = (DiskShare) session.connectShare(share_name)
    }

    void close() {
        share.close()
        session.close()
        connection.close()
    }

**Comment by TheRealHypo** [comment]
If I call share.close() on a Single instance of this object, it won't close the share instances on the other instances of this obejct I have. 

So e.g. 
 a.getShare().close() would not close b.getShare()

a.getConnection().close() would close a.getShare(), but b.getShare() as well.

**Comment by ecki** [comment]
That asumption might not be true, I see that connectShare returns cached instances. Maybe you can print their idendityHash or check for same object to verify: https://github.com/hierynomus/smbj/blob/master/src/main/java/com/hierynomus/smbj/session/Session.java#L135

**Comment by TheRealHypo** [comment]
I've tested it earlier though via Evaluator. If I'm back at work tomorrow, I'll write a Test for that specific case. I'm pretty sure that I could use the share on other Instances, until I closed one of the other handles.

**Comment by hierynomus** [comment]
A single client will try to reuse as much as possible, as it is specified in the spec to do so. So multiple sessions will be multiplexed over a single connection as long as they reside on the same target server. Same for multiple share opens by the same user, they will reuse the session, etc. 

If you reuse the same client, only close the File's, but not the session/share. And only close the client (which will close all sub-objects) when you're done. Or create a new client per request and not reuse it across threads.

**Comment by TheRealHypo** [comment]
So only close the DiskEntry not the Share, Session or Connection ?

**Comment by pepijnve** [comment]
> So only close the DiskEntry not the Share, Session or Connection ?

Indeed. In general, you don't want to create a new unique connection, session or share (tree connect) per file. Doing that would be pretty inefficient. There's quite some work involved in setting up and tearing down each of these.

**Comment by TheRealHypo** [comment]
Thanks a lot then, I'll look now how good I understand that, and try to implement it. 

What you are saying makes totally sense, but I never encountered a TreeConnect before (I'm just a 7th semester student), so I didn't really knew how to read the code in the back. Some Documentation would be perfect, to clear such things up 🙊

**Comment by pepijnve** [comment]
The SMB specification is your friend 😄 See https://msdn.microsoft.com/en-us/library/cc246793.aspx for instance.

`TREE_CONNECT` is the name of the request in the specification that 'opens' the share.

**Comment by felixRajkumar** [comment]
> > So only close the DiskEntry not the Share, Session or Connection ?

**Only close the entry** does it implies the to close the DiskShare or something else

**Comment by felixRajkumar** [comment]
> So only close the DiskEntry not the Share, Session or Connection ?

Here I would like to know whether the term DiskEntry specifies the DiskShare in the following code.

` Connection connection = client.connect("SERVERNAME");
        AuthenticationContext ac = new AuthenticationContext("USERNAME", "PASSWORD".toCharArray(), "DOMAIN");
        Session session = connection.authenticate(ac);
        // Connect to Share
        DiskShare share = (DiskShare) session.connectShare("SHARENAME");
            //Code for reading 
       `

**Comment by hierynomus** [comment]
`entry` references to a `DiskEntry`, so a `File` or `Directory`

**Comment by felixRajkumar** [comment]
> `entry` references to a `DiskEntry`, so a `File` or `Directory`

Will I get any issues if I didn't close the diskShare. I have asked a similar question here #543
